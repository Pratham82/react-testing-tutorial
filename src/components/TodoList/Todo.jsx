import React from 'react'
export default function Todo({ id, title, status }) {
  const TodoStyles = {
    div: {
      padding: '10px',
    },
    span: {
      fontWeight: 'bold',
    },
    statuses: {
      completed: {
        color: 'green',
        textDecoration: 'line-through',
      },
      pending: {
        color: 'red',
      },
    },
  }
  return (
    <div style={TodoStyles.div} data-testid={`todo-${id}`}>
      &nbsp;
      <span
        style={
          status ? TodoStyles.statuses.completed : TodoStyles.statuses.pending
        }
      >
        {status ? <strike>{title}</strike> : title}
      </span>
    </div>
  )
}
