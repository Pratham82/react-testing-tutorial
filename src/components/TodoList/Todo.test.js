import { render, screen, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import Todo from './Todo'

afterEach(() => {
  cleanup()
})

// Check if first todo is correct and not empty
test('should render todo component', () => {
  const { id, title, status } = { id: 1, title: 'Do Laundry', status: false }
  render(<Todo id={id} title={title} status={status} />)
  const todoElement = screen.getByTestId('todo-1')
  expect(todoElement).toBeInTheDocument()
  // Check if it is exact todo
  expect(todoElement).toHaveTextContent('Do Laundry')
  // Check if empty
  expect(todoElement).toHaveTextContent(/[A-Z]/gi)
})

// Check completed Todo
test('should render completed todo item', () => {
  const { id, title, status } = { id: 3, title: 'Work out', status: true }
  render(<Todo id={id} title={title} status={status} />)
  const completedTodo = screen.getByTestId('todo-3')
  expect(completedTodo).toContainHTML('strike')
})

// Match snapshot
test('matches snapshot', () => {
  const { id, title, status } = { id: 1, title: 'Do Laundry', status: false }
  //* 1. create a tree of the component (contains the object form of our component)
  const tree = renderer
    .create(<Todo id={id} title={title} status={status} />)
    .toJSON()

  //* 2. Match our tree with the snapshot
  //* This will match the component structure and returns error if there is any change in the structure of the tree
  expect(tree).toMatchSnapshot()
})
