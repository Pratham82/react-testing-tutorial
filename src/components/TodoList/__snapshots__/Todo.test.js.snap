// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="todo-1"
  style={
    Object {
      "padding": "10px",
    }
  }
>
  Â 
  <span
    style={
      Object {
        "color": "red",
      }
    }
  >
    Do Laundry
  </span>
</div>
`;
